File: sol/983/983.py
from bisect import bisect_left

class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        n = len(days)
        memo = [0] * (n + 1)
        for i in range(n):
            memo[i + 1] = min(
                memo[i] + costs[0],
                memo[bisect_left(days, days[i] - 6, hi=i)] + costs[1],
                memo[bisect_left(days, days[i] - 29, hi=i)] + costs[2]
            )
        return memo[n]

File: sol/983/983.cpp
#include <algorithm>

struct Solution {
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        int n = days.size();
        vector<int> memo(n + 1);
        memo[0] = 0;
        for (int i = 0; i < n; ++i){
            memo[i + 1] = std::min({
                memo[i] + costs[0],
                memo[std::lower_bound(days.begin(), days.begin() + i, days[i] - 6) - days.begin()] + costs[1],
                memo[std::lower_bound(days.begin(), days.begin() + i, days[i] - 29) - days.begin()] + costs[2]
            });
        }
        return memo[n];
    }
};
